bin/grade_item.py

```diff
@@ -0,0 +1,412 @@
+    test_output_path = os.path.join(SUBMITTY_DATA_DIR,"courses",obj["semester"],obj["course"],"test_output",obj["gradeable"])
+    bin_path = os.path.join(SUBMITTY_DATA_DIR,"courses",obj["semester"],obj["course"],"bin")
+
+    #checkout_path="$SUBMITTY_DATA_DIR/courses/$semester/$course/checkout/$gradeable/$who/$version"
+
+    results_path = os.path.join(SUBMITTY_DATA_DIR,"courses",obj["semester"],obj["course"],"results",obj["gradeable"],obj["who"],str(obj["version"]))
+
+    # grab a copy of the current results_history.json file (if it exists)
+    global_results_history_file_location=os.path.join(results_path,"results_history.json")
+    # FIXME - HISTORY FILE NEEDS WORK
+
+    # --------------------------------------------------------------------
+    # MAKE TEMPORARY DIRECTORY & COPY THE NECESSARY FILES THERE
+    tmp=tempfile.mkdtemp()
+
+    # switch to tmp directory
+    os.chdir(tmp)
+
+    # make the logs directory
+    tmp_logs = os.path.join(tmp,"tmp_logs")
+    os.makedirs(tmp_logs)
+
+    # grab the submission time
+    with open (os.path.join(submission_path,".submit.timestamp")) as submission_time_file:
+        submission_time=submission_time_file.read()
+
+    
+    # --------------------------------------------------------------------
+    # COMPILE THE SUBMITTED CODE
+
+    with open(os.path.join(tmp_logs,"overall.txt"),'a') as f:
+        print ("====================================\nCOMPILATION STARTS", file=f)
+    
+    # copy submitted files to the tmp compilation directory
+    tmp_compilation = os.path.join(tmp,"TMP_COMPILATION")
+    os.mkdir(tmp_compilation)
+    os.chdir(tmp_compilation)
+    
+    # get info from the gradeable config file
+    json_config=os.path.join(SUBMITTY_DATA_DIR,"courses",obj["semester"],obj["course"],"config","form","form_"+obj["gradeable"]+".json")
+    with open(json_config, 'r') as infile:
+        gradeable_config_obj = json.load(infile)
+
+    # get info from the gradeable config file
+    complete_config=os.path.join(SUBMITTY_DATA_DIR,"courses",obj["semester"],obj["course"],"config","complete_config","complete_config_"+obj["gradeable"]+".json")
+    with open(complete_config, 'r') as infile:
+        complete_config_obj = json.load(infile)
+
+    gradeable_upload_type=gradeable_config_obj["upload_type"]
+    #print ("UPLOAD TYPE ",gradeable_upload_type)
+    #FIXME:  deal with svn/git/whatever
+
+    gradeable_deadline = gradeable_config_obj["date_due"]
+
+    patterns_submission_to_compilation = complete_config_obj["autograding"]["submission_to_compilation"]
+    pattern_copy("submission_to_compilation",patterns_submission_to_compilation,submission_path,tmp_compilation,tmp_logs)
+    
+    # copy any instructor provided code files to tmp compilation directory
+    copy_contents_into(test_code_path,tmp_compilation)
+
+    subprocess.call("ls -la "+tmp_compilation+" >> "+tmp_logs+"/overall.txt",shell=True)
```

Comment:
Is there a compelling reason to have `shell=True` here? I'd rather see `subprocess.call(['ls', '-la', tmp_compilation], stdout=file(tmp_logs + "/overall.txt", 'a')` as using the shell is not great for portability.

Repeat this for the several other times you do this.